File: src/codechicken/multipart/scalatraits/TModelRenderTile.scala
package codechicken.multipart.scalatraits

import codechicken.lib.render.CCRenderState
import codechicken.lib.vec.Vector3
import codechicken.multipart._
import net.minecraft.block.state.IBlockState
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.block.model.IBakedModel
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.util.BlockRenderLayer
import net.minecraftforge.client.ForgeHooksClient

import scala.collection.mutable.ListBuffer

/**
  * Mixin implementation for IModelRenderPart.
  */
trait TModelRenderTile extends TileMultipartClient
{
    private var modelPartList = ListBuffer[IModelRenderPart]()

    override def copyFrom(that:TileMultipart)
    {
        super.copyFrom(that)
        that match {
            case mt:TModelRenderTile => modelPartList = mt.modelPartList
            case _ =>
        }
    }

    override def bindPart(part:TMultiPart)
    {
        super.bindPart(part)
        part match {
            case mp:IModelRenderPart => modelPartList += mp
            case _ =>
        }
    }

    override def clearParts()
    {
        super.clearParts()
        modelPartList.clear()
    }

    override def partRemoved(part:TMultiPart, p:Int)
    {
        super.partRemoved(part, p)
        part match {
            case mp:IModelRenderPart => modelPartList -= mp
            case _ =>
        }
    }

    override def renderStatic(pos:Vector3, layer:BlockRenderLayer) =
    {
        var r = super.renderStatic(pos, layer)

        renderModel(modelPartList.filter(_.canRenderInLayer(layer)), {(model, state) =>
            Minecraft.getMinecraft.getBlockRendererDispatcher.getBlockModelRenderer
                    .renderModel(getWorld, model, state, getPos, CCRenderState.getBuffer, true)
            r |= true
        })

        r
    }

    override def renderDamage(pos:Vector3, texture:TextureAtlasSprite)
    {
        super.renderDamage(pos, texture)

        Minecraft.getMinecraft.objectMouseOver match {
            case hit:PartRayTraceResult => partList(hit.partIndex) match {
                case p:IModelRenderPart =>
                    renderModel(ListBuffer(p), {(model, state) =>
                        val dm = ForgeHooksClient.getDamageModel(model, texture, state, getWorld, getPos)
                        Minecraft.getMinecraft.getBlockRendererDispatcher.getBlockModelRenderer
                                .renderModel(getWorld, dm, state, getPos, CCRenderState.getBuffer, true)
                    })
                case _ =>
            }
            case _ =>
        }
    }

    private def renderModel(list:ListBuffer[IModelRenderPart], f:(IBakedModel, IBlockState) => Unit)
    {
        import MultiPartRegistryClient._
        for (mp <- list) {
            val container = getModelPartContainer(mp)
            val state = mp.getCurrentState(container.getBaseState)
            val disp = Minecraft.getMinecraft.getBlockRendererDispatcher

            disp.getBlockModelShapes.getModelForState(state) match {
                case null =>
                case model => f(model, state)
            }
        }
    }
}

File: src/main/scala/codechicken/multipart/trait/TModelRenderTile.java
//package codechicken.multipart.trait;
//
//import codechicken.lib.render.CCRenderState;
//import codechicken.lib.vec.Vector3;
//import codechicken.mixin.forge.TraitSide;
//import codechicken.multipart.TileMultipart;
//import codechicken.multipart.TileMultipartClient;
//import codechicken.multipart.api.annotation.MultiPartTrait;
//import codechicken.multipart.api.part.IModelRenderPart;
//import codechicken.multipart.api.part.TMultiPart;
//import com.mojang.blaze3d.matrix.MatrixStack;
//import net.minecraft.client.Minecraft;
//import net.minecraft.client.renderer.BlockRendererDispatcher;
//import net.minecraft.client.renderer.RenderType;
//import net.minecraft.world.ILightReader;
//
//import java.util.LinkedList;
//import java.util.List;
//import java.util.Random;
//
///**
// * Created by covers1624 on 1/9/20.
// */
////@MultiPartTrait (value = IModelRenderPart.class, side = TraitSide.CLIENT)
//class TModelRenderTile extends TileMultipart implements TileMultipartClient {
//
//    private final List<IModelRenderPart> modelPartList = new LinkedList<>();
//
//    @Override
//    public void copyFrom(TileMultipart that) {
//        super.copyFrom(that);
//        if (that instanceof TModelRenderTile) {
//            modelPartList.clear();
//            modelPartList.addAll(((TModelRenderTile) that).modelPartList);
//        }
//    }
//
//    @Override
//    public void bindPart(TMultiPart part) {
//        super.bindPart(part);
//        if (part instanceof IModelRenderPart) {
//            modelPartList.add((IModelRenderPart) part);
//        }
//    }
//
//    @Override
//    public void partRemoved(TMultiPart part, int p) {
//        super.partRemoved(part, p);
//        if (part instanceof IModelRenderPart) {
//            modelPartList.remove(part);
//        }
//    }
//
//    @Override
//    public void clearParts() {
//        super.clearParts();
//        modelPartList.clear();
//    }
//
//    @Override
//    public boolean renderStatic(RenderType layer, CCRenderState ccrs) {
//        boolean ret = TileMultipartClient.super.renderStatic(layer, ccrs);
//        BlockRendererDispatcher rendererDispatcher = Minecraft.getInstance().getBlockRendererDispatcher();
//
//        ILightReader world = ccrs.lightMatrix.access;
//
//        MatrixStack stack = new MatrixStack();
//        for (IModelRenderPart part : modelPartList) {
//            Random random = new Random();
//            if (part.canRenderInLayer(layer)) {
//                stack.push();
//                ret |= rendererDispatcher.renderModel(part.getCurrentState(), getPos(), world, stack, ccrs.getConsumer(), true, random, part.getModelData());
//                stack.pop();
//            }
//        }
//        return ret;
//    }
//}
